File: docs/generate.py
# Copyright 2021, 2022, 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Helper binary to convert Stardoc output into Org Mode format."""

import argparse
from collections.abc import Callable
import html.parser
import io
import pathlib
import re
import textwrap
from typing import Optional
import urllib.parse

import commonmark
import commonmark.node
import commonmark.render.renderer

from docs import stardoc_output_pb2

# The generated protocol buffer modules generate their members dynamically, so
# Pylint can’t find them.
# pylint: disable=no-member

def main() -> None:
    """Main function."""
    parser = argparse.ArgumentParser(allow_abbrev=False)
    parser.add_argument('input', type=pathlib.Path)
    parser.add_argument('output', type=pathlib.Path)
    opts = parser.parse_args()
    module = stardoc_output_pb2.ModuleInfo.FromString(opts.input.read_bytes())
    # Force Unix-style line endings for consistent results.  See
    # https://github.com/bazelbuild/stardoc/issues/110.
    with opts.output.open(mode='xt', encoding='utf-8', newline='\n') as file:
        generator = _Generator(file)
        generator.run(module)


class _Generator:

    _ATTRIBUTE_TYPE = {
        stardoc_output_pb2.NAME: 'Name',
        stardoc_output_pb2.INT: 'Integer',
        stardoc_output_pb2.LABEL: 'Label',
        stardoc_output_pb2.STRING: 'String',
        stardoc_output_pb2.STRING_LIST: 'List of strings',
        stardoc_output_pb2.INT_LIST: 'List of integers',
        stardoc_output_pb2.LABEL_LIST: 'List of labels',
        stardoc_output_pb2.BOOLEAN: 'Boolean',
        stardoc_output_pb2.LABEL_STRING_DICT: 'Dictionary string → label',
        stardoc_output_pb2.STRING_DICT: 'Dictionary string → string',
        stardoc_output_pb2.STRING_LIST_DICT:
            'Dictionary string → list of strings',
        stardoc_output_pb2.OUTPUT: 'Output file',
        stardoc_output_pb2.OUTPUT_LIST: 'List of output files',
    }

    _MANDATORY = {
        False: 'optional',
        True: 'mandatory',
    }

    def __init__(self, file: io.TextIOBase):
        self._file = file

    def run(self, module: stardoc_output_pb2.ModuleInfo) -> None:
        """Writes the generated Org Mode output."""
        self._write(_markdown(module.module_docstring))
        for rule in module.rule_info:
            self._rule(rule)
        for provider in module.provider_info:
            self._provider(provider)
        for func in module.func_info:
            self._function(func)
        for aspect in module.aspect_info:
            self._aspect(aspect)
        for extension in module.module_extension_info:
            self._extension(extension)
        for rule in module.repository_rule_info:
            self._repo_rule(rule)

    def _rule(self, rule: stardoc_output_pb2.RuleInfo) -> None:
        name = rule.rule_name
        attrs = ', '.join(a.name if a.mandatory else f'[{a.name}]'
                          for a in rule.attribute)
        self._write(f'#+ATTR_TEXINFO: :options Rule {name} ({attrs})\n')
        self._write('#+BEGIN_deffn\n')
        self._write(_markdown(rule.doc_string).lstrip())
        self._write(f'The ~{name}~ rule supports the following attributes:\n\n')
        for attr in rule.attribute:
            self._attribute(attr)
        self._write('#+END_deffn\n\n')

    def _function(self, func: stardoc_output_pb2.StarlarkFunctionInfo) -> None:
        name = func.function_name
        params = ', '.join(p.name if p.mandatory else f'[{p.name}]'
                           for p in func.parameter)
        self._write(f'#+ATTR_TEXINFO: :options {name} ({params})\n')
        self._write('#+BEGIN_defun\n')
        self._write(_markdown(func.doc_string).lstrip())
        for param in func.parameter:
            self._parameter(param)
        returns = getattr(func, 'return').doc_string
        if returns:
            self._write(f'Returns: {returns}\n\n')
        if func.deprecated.doc_string:
            raise ValueError(f'unsupported deprecated function {name}')
        self._write('#+END_defun\n\n')

    def _parameter(self, param: stardoc_output_pb2.FunctionParamInfo) -> None:
        doc = _markdown(param.doc_string).strip()
        if not doc.endswith('.'):
            raise ValueError(
                f'documentation string {doc!r} should end with a period')
        suffixes = [self._MANDATORY[param.mandatory]]
        if param.default_value:
            suffixes.append(f'default: ~{param.default_value}~')
        suffix = '; '.join(suffixes)
        self._item(f'{param.name} :: {doc}.  {suffix}.')

    def _provider(self, provider: stardoc_output_pb2.ProviderInfo) -> None:
        name = provider.provider_name
        fields = ' '.join(f.name for f in provider.field_info)
        self._write(f'#+ATTR_TEXINFO: :options Provider {name} {fields}\n')
        self._write('#+BEGIN_deftp\n')
        self._write(_markdown(provider.doc_string).lstrip())
        self._write(f'The ~{name}~ provider has the following fields:\n\n')
        for field in provider.field_info:
            doc = _markdown(field.doc_string).strip()
            if not doc.endswith('.'):
                raise ValueError(
                    f'documentation string {doc!r} should end with a period')
            self._item(f'~{field.name}~ :: {doc}')
        self._write('#+END_deftp\n\n')

    def _aspect(self, aspect: stardoc_output_pb2.AspectInfo) -> None:
        name = aspect.aspect_name
        attrs = ' '.join(a.name if a.mandatory else f'[{a.name}]'
                         for a in aspect.attribute)
        self._write(f'#+ATTR_TEXINFO: :options Aspect {name} {attrs}\n')
        self._write('#+BEGIN_deffn\n')
        self._write(_markdown(aspect.doc_string).lstrip())
        if aspect.aspect_attribute:
            attrs = ', '.join(f'~{a}~' for a in aspect.aspect_attribute)
            self._write(f'This aspect propagates along the following '
                        f'attributes: {attrs}\n')
        for attr in aspect.attribute:
            self._attribute(attr)
        self._write('#+END_deffn\n\n')

    def _extension(self, ext: stardoc_output_pb2.ModuleExtensionInfo) -> None:
        name = ext.extension_name
        tags = ' '.join(t.tag_name for t in ext.tag_class)
        self._write(
            f'#+ATTR_TEXINFO: :options {{Module extension}} {name} {tags}\n')
        self._write('#+BEGIN_deftp\n')
        self._write(_markdown(ext.doc_string).lstrip())
        self._write(f'The ~{name}~ module extension '
                    f'provides the following tag classes:\n\n')
        for tag in ext.tag_class:
            self._tag_class(name, tag)
        self._write('#+END_deftp\n\n')

    def _tag_class(self, extension_name: str,
                   tag: stardoc_output_pb2.ModuleExtensionTagClassInfo) -> None:
        name = tag.tag_name
        attrs = ', '.join(a.name if a.mandatory else f'[{a.name}]'
                          for a in tag.attribute)
        self._write(f'#+ATTR_TEXINFO: :options '
                    f'{{Tag class}} {extension_name} {name} ({attrs})\n')
        self._write('#+BEGIN_defop\n')
        self._write(_markdown(tag.doc_string).lstrip())
        self._write(
            f'The ~{name}~ tag class supports the following attributes:\n\n')
        for attr in tag.attribute:
            self._attribute(attr)
        self._write('#+END_defop\n\n')

    def _repo_rule(self, rule: stardoc_output_pb2.RepositoryRuleInfo) -> None:
        name = rule.rule_name
        attrs = ', '.join(a.name if a.mandatory else f'[{a.name}]'
                          for a in rule.attribute)
        self._write(
            f'#+ATTR_TEXINFO: :options {{Repository rule}} {name} ({attrs})\n')
        self._write('#+BEGIN_deffn\n')
        self._write(_markdown(rule.doc_string).lstrip())
        self._write(f'The ~{name}~ repository rule '
                    f'supports the following attributes:\n\n')
        for attr in rule.attribute:
            self._attribute(attr)
        if rule.environ:
            env = ', '.join(f'~{e}~' for e in rule.environ)
            self._write(f'It depends on the following '
                        f'environment variables: {env}\n\n')
        self._write('#+END_deffn\n\n')

    def _attribute(self, attr: stardoc_output_pb2.AttributeInfo) -> None:
        doc = _markdown(attr.doc_string).strip()
        if not doc.endswith('.'):
            raise ValueError(
                f'documentation string {doc!r} should end with a period')
        suffixes = [self._ATTRIBUTE_TYPE[attr.type],
                    self._MANDATORY[attr.mandatory]]
        if attr.default_value:
            suffixes.append(f'default: ~{attr.default_value}~')
        if attr.provider_name_group:
            (group,) = attr.provider_name_group
            names = ', '.join(f'~{name}~' for name in group.provider_name)
            suffixes.append(f'required providers: {names}')
        suffix = '; '.join(suffixes)
        self._item(f'~{attr.name}~ :: {doc}  {suffix}.')

    def _item(self, text: str) -> None:
        self._write(
            _fill(text, initial_indent='- ', subsequent_indent='  ') + '\n')

    def _write(self, text: str) -> None:
        self._file.write(text)


def _markdown(text: str) -> str:
    """Convert a Markdown snippet to Org-mode."""
    # Bazel (including Stardoc) interprets all files as Latin-1,
    # cf. https://bazel.build/concepts/build-files.  However, our files all use
    # UTF-8, leading to double encoding.  Reverse that effect here.
    text = text.strip().encode('latin-1').decode('utf-8')
    document = commonmark.Parser().parse(text)
    text = _OrgRenderer().render(document)
    return text + '\n'


def _fill(text: str, *,
          initial_indent: str = '', subsequent_indent: str = '') -> str:
    return textwrap.fill(
        text, width=80, tabsize=4, fix_sentence_endings=True,
        initial_indent=initial_indent, subsequent_indent=subsequent_indent,
        break_long_words=False, break_on_hyphens=False)


class _OrgRenderer(commonmark.render.renderer.Renderer):

    _LANGUAGE = {'bash': 'sh'}

    def __init__(self):
        super().__init__()
        self._indent = ''

    def out(self, s) -> None:
        indent = self._indent if self.last_out.endswith('\n') else ''
        self.lit(indent + self._escape(s))

    @staticmethod
    def _escape(string: str) -> str:
        # See https://orgmode.org/manual/Escape-Character.html.
        return re.sub(r'([\[\]*/_=~+])', '\\1\N{ZERO WIDTH SPACE}', string)

    # The interface is mandated by CommonMark.  Silence false-positive lint
    # errors.
    # pylint: disable=missing-function-docstring, unused-argument

    def document(self, node: commonmark.node.Node, entering: bool) -> None:
        self.cr()

    def text(self, node: commonmark.node.Node, entering: bool) -> None:
        self.out(node.literal)

    def softbreak(self, node: commonmark.node.Node, entering: bool) -> None:
        self.cr()

    def linebreak(self, node: commonmark.node.Node, entering: bool) -> None:
        self.lit(r'\\')
        self.cr()

    def paragraph(self, node: commonmark.node.Node, entering: bool) -> None:
        if node.parent.t != 'item':
            self.lit('\n')

    def list(self, node: commonmark.node.Node, entering: bool) -> None:
        if entering:
            self.cr()

    def item(self, node: commonmark.node.Node, entering: bool) -> None:
        if entering:
            assert not self._indent  # no support for nested lists
            self.lit('- ')
            self._indent = '  '
        else:
            assert self._indent == '  '  # no support for nested lists
            self._indent = ''
            self.cr()

    def emph(self, node: commonmark.node.Node, entering: bool) -> None:
        self.lit('/')

    def code(self, node: commonmark.node.Node, entering: bool) -> None:
        self.lit(f'~{node.literal}~')

    def code_block(self, node: commonmark.node.Node, entering: bool) -> None:
        lang = self._LANGUAGE[node.info]
        self.lit(f'#+BEGIN_SRC {lang}\n{node.literal}#+END_SRC\n\n')
        self.lit('#+TEXINFO: @noindent')

    def link(self, node: commonmark.node.Node, entering: bool) -> None:
        if entering:
            dest = node.destination
            # CommonMark helpfully URL-escapes link destinations, but this
            # prevents links to Info nodes containing spaces.
            match = re.fullmatch(r'(info:[^#:]+[#:])(.*%.*)', dest)
            if match:
                dest = match[1] + urllib.parse.unquote(match[2])
            self.lit(f'[[{dest}][')
        else:
            self.lit(']]')

    def html_inline(self, node: commonmark.node.Node, entering: bool) -> None:
        writer = io.StringIO()
        parser = _HTMLParser(writer)
        parser.feed(node.literal)
        self.lit(writer.getvalue())

    # Signal an error if we don’t implement something.
    def __getattr__(self, name) -> Callable[[commonmark.node.Node, bool], None]:
        if not name or name.startswith('_'):
            raise AttributeError(name)
        # Assume this is supposed to be a rendering method.
        return self._unknown

    def _unknown(self, node: commonmark.node.Node, entering: bool) -> None:
        raise NotImplementedError(f'unknown node type {node.t!r}')


class _HTMLParser(html.parser.HTMLParser):

    _TAGS = {
        'p': ('\n\n', ''),
        'code': ('@@texinfo:@code{@@', '@@texinfo:}@@'),
        'var': ('@@texinfo:@var{@@', '@@texinfo:}@@'),
    }

    def __init__(self, writer: io.TextIOBase):
        super().__init__()
        self._writer = writer

    def handle_starttag(self, tag: str,
                        attrs: list[tuple[str, Optional[str]]]) -> None:
        if attrs:
            raise NotImplementedError(f'got attributes {attrs} for <{tag}> tag')
        start, _ = self._TAGS[tag]
        self._writer.write(start)

    def handle_endtag(self, tag: str) -> None:
        _, end = self._TAGS[tag]
        self._writer.write(end)

    def handle_data(self, data: str) -> None:
        self._writer.write(data)


if __name__ == '__main__':
    main()

File: internal/stardoc2org/generate.go
// Copyright 2021-2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Helper binary to convert Stardoc output into Org Mode format.
package main

import (
	"errors"
	"flag"
	"fmt"
	"io"
	"log"
	"net/url"
	"os"
	"regexp"
	"runtime/debug"
	"strings"
	"text/template"
	"unicode"
	"unicode/utf8"

	_ "embed"

	"github.com/yuin/goldmark/v2/ast"
	"github.com/yuin/goldmark/v2/parser"
	"github.com/yuin/goldmark/v2/renderer"
	"golang.org/x/text/cases"
	"golang.org/x/text/language"
	"google.golang.org/protobuf/proto"

	spb "github.com/phst/rules_elisp/internal/stardoc_output_go_proto"
)

// Main function.
func main() {
	flag.Usage = usage
	flag.Parse()
	if flag.NArg() != 2 {
		usage()
		os.Exit(2)
	}
	input := flag.Arg(0)
	output := flag.Arg(1)
	b, err := os.ReadFile(input)
	if err != nil {
		log.Fatal(err)
	}
	var module spb.ModuleInfo
	if err := proto.Unmarshal(b, &module); err != nil {
		log.Fatal(err)
	}
	file, err := os.OpenFile(output, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0400)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()
	generator := newGenerator(file)
	if err := generator.run(&module); err != nil {
		log.Fatal(err)
	}
	if err := file.Sync(); err != nil {
		log.Fatal(err)
	}
	if err := file.Close(); err != nil {
		log.Fatal(err)
	}
}

func usage() {
	fmt.Fprintln(os.Stderr, "Usage: generate INPUT OUTPUT")
	flag.PrintDefaults()
}

type generator struct {
	file io.Writer
}

var attributeType = map[spb.AttributeType]string{
	spb.AttributeType_NAME:              "Name",
	spb.AttributeType_INT:               "Integer",
	spb.AttributeType_LABEL:             "Label",
	spb.AttributeType_STRING:            "String",
	spb.AttributeType_STRING_LIST:       "List of strings",
	spb.AttributeType_INT_LIST:          "List of integers",
	spb.AttributeType_LABEL_LIST:        "List of labels",
	spb.AttributeType_BOOLEAN:           "Boolean",
	spb.AttributeType_LABEL_STRING_DICT: "Dictionary string → label",
	spb.AttributeType_STRING_DICT:       "Dictionary string → string",
	spb.AttributeType_STRING_LIST_DICT:  "Dictionary string → list of strings",
	spb.AttributeType_OUTPUT:            "Output file",
	spb.AttributeType_OUTPUT_LIST:       "List of output files",
}

func formatAttributeType(t spb.AttributeType) (string, error) {
	s, ok := attributeType[t]
	if !ok {
		return "", fmt.Errorf("unknown attribute type %s", t)
	}
	return s, nil
}

var mandatory = map[bool]string{
	false: "optional",
	true:  "mandatory",
}

func formatMandatory(b bool) string {
	return mandatory[b]
}

func newGenerator(file io.Writer) *generator {
	return &generator{file}
}

// Writes the generated Org Mode output.
func (g *generator) run(module *spb.ModuleInfo) (err error) {
	defer runRecover(&err)
	g.doRun(module)
	return
}

func runRecover(err *error) {
	switch x := recover().(type) {
	case nil:
		return
	case error:
		*err = fmt.Errorf("Generator panic: %w", x)
	default:
		*err = fmt.Errorf("Generator panic: %#v", x)
	}
	debug.PrintStack()
}

func (g *generator) doRun(module *spb.ModuleInfo) {
	tpl := template.New("reference.org.template")
	item := func(template string, data any) (string, error) {
		var w strings.Builder
		err := tpl.ExecuteTemplate(&w, template, data)
		if err != nil {
			return "", err
		}
		s := w.String()
		if s == "" {
			return "", nil
		}
		return fill(s, "- ", "  ") + "\n", nil
	}
	funcs := template.FuncMap{
		"containsRune":    strings.ContainsRune,
		"hasPrefix":       strings.HasPrefix,
		"trimSpace":       strings.TrimSpace,
		"lstrip":          lstrip,
		"capitalize":      capitalize,
		"mandatory":       formatMandatory,
		"attributeType":   formatAttributeType,
		"requireEmpty":    requireEmpty,
		"requireNonEmpty": requireNonEmpty,
		"requirePeriod":   requirePeriod,
		"exactlyOne":      exactlyOne,
		"markdown":        markdown,
		"item":            item,
	}
	tpl = template.Must(tpl.Funcs(funcs).Parse(templateText))
	if err := tpl.Execute(g.file, module); err != nil {
		panic(err)
	}
}

//go:embed reference.org.template
var templateText string

// Convert a Markdown snippet to Org-mode.
func markdown(text string) string {
	text = strings.TrimSpace(text)
	if text == "" {
		panic(errors.New("Missing docstring"))
	}
	source := []byte(text)
	doc := parser.New().Parse(source)
	renderer := newRenderer()
	var w strings.Builder
	if err := renderer.Render(&w, source, doc); err != nil {
		panic(err)
	}
	return w.String() + "\n"
}

func fill(text, initialIndent, subsequentIndent string) string {
	text = regexp.MustCompile(`([.!?])\n+`).ReplaceAllString(text, "$1  ")
	text = strings.ReplaceAll(text, "\t", "    ")
	words := regexp.MustCompile(`[ \n]`).Split(text, -1)
	var b strings.Builder
	b.WriteString(initialIndent)
	i := 0
	width := 80 - len(initialIndent)
	for _, w := range words {
		l := utf8.RuneCountInString(w)
		if i+l >= width {
			b.WriteByte('\n')
			b.WriteString(subsequentIndent)
			i = 0
		} else if i > 0 {
			b.WriteByte(' ')
			i++
		}
		b.WriteString(w)
		i += l
	}
	return b.String()
}

func lstrip(s string) string {
	return strings.TrimLeftFunc(s, unicode.IsSpace)
}

func capitalize(s string) string {
	return cases.Title(language.English, cases.NoLower).String(s)
}

func requireEmpty(s string) (string, error) {
	if s != "" {
		return "", fmt.Errorf("string %q isn’t empty", s)
	}
	return "", nil
}

func requireNonEmpty(s string) (string, error) {
	if s == "" {
		return s, errors.New("empty string")
	}
	return s, nil
}

func requirePeriod(s string) (string, error) {
	if !strings.HasSuffix(s, ".") {
		return "", fmt.Errorf("documentation string %q should end with a period", s)
	}
	return s, nil
}

func exactlyOne(groups []*spb.ProviderNameGroup) (*spb.ProviderNameGroup, error) {
	if n := len(groups); n != 1 {
		return nil, fmt.Errorf("got %d provider name groups, want one", n)
	}
	return groups[0], nil
}

type markdownRenderer struct {
	helper *renderer.Helper[io.Writer, rendererConfig]
}

func newRenderer() *markdownRenderer {
	orgRenderer := newOrgRenderer()
	helper := new(renderer.HelperBuilder[io.Writer, rendererConfig]).Options(
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindDocument, renderer.NodeRendererFunc(orgRenderer.document)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindText, renderer.NodeRendererFunc(orgRenderer.text)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindParagraph, renderer.NodeRendererFunc(orgRenderer.paragraph)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindList, renderer.NodeRendererFunc(orgRenderer.list)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindListItem, renderer.NodeRendererFunc(orgRenderer.item)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindEmphasis, renderer.NodeRendererFunc(orgRenderer.emph)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindCodeSpan, renderer.NodeRendererFunc(orgRenderer.code)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindCodeBlock, renderer.NodeRendererFunc(orgRenderer.codeBlock)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindLink, renderer.NodeRendererFunc(orgRenderer.link)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindRawHTML, renderer.NodeRendererFunc(orgRenderer.htmlInline)),

		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindAutoLink, renderer.NodeRendererFunc(orgRenderer.unknown)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindBlockquote, renderer.NodeRendererFunc(orgRenderer.unknown)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindHTMLBlock, renderer.NodeRendererFunc(orgRenderer.unknown)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindHeading, renderer.NodeRendererFunc(orgRenderer.unknown)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindImage, renderer.NodeRendererFunc(orgRenderer.unknown)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindLinkReferenceDefinition, renderer.NodeRendererFunc(orgRenderer.unknown)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindStrong, renderer.NodeRendererFunc(orgRenderer.unknown)),
		renderer.WithNodeRenderer[io.Writer, rendererConfig](ast.KindThematicBreak, renderer.NodeRendererFunc(orgRenderer.unknown)),
	).Build()
	return &markdownRenderer{helper}
}

func (r *markdownRenderer) Render(w io.Writer, source []byte, n ast.Node, opts ...renderer.RenderOption) error {
	return r.helper.Render(w, source, n, opts...)
}

func (r *markdownRenderer) RenderStringSource(w io.Writer, source string, n ast.Node, opts ...renderer.RenderOption) error {
	return r.helper.RenderStringSource(w, source, n, opts...)
}

var _ renderer.Renderer[io.Writer] = (*markdownRenderer)(nil)

type rendererConfig struct {
	Config renderer.Config[io.Writer, rendererConfig]
}

type orgRenderer struct {
	indent  string
	lastOut string
}

var rendererLanguage = map[string]string{
	"sh": "sh",
	"c":  "c",
}

func newOrgRenderer() *orgRenderer {
	return &orgRenderer{"", ""}
}

func (r *orgRenderer) lit(w io.Writer, s string) {
	if _, err := io.WriteString(w, s); err != nil {
		panic(err)
	}
	r.lastOut = s
}

func (r *orgRenderer) cr(w io.Writer) {
	if r.lastOut != "\n" {
		r.lit(w, "\n")
	}
}

func (r *orgRenderer) document(writer io.Writer, source []byte, n ast.Node, entering bool, rc renderer.Context) (ast.WalkStatus, error) {
	_ = n.(*ast.Document)
	r.cr(writer)
	return ast.WalkContinue, nil
}

func (r *orgRenderer) text(writer io.Writer, source []byte, n ast.Node, entering bool, rc renderer.Context) (ast.WalkStatus, error) {
	node := n.(*ast.Text)
	if entering {
		s := node.Value.Str(source)
		s = regexp.MustCompile(`\\(.)`).ReplaceAllString(s, "$1")
		indent := ""
		if strings.HasSuffix(r.lastOut, "\n") {
			indent = r.indent
		}
		// See https://orgmode.org/manual/Escape-Character.html.
		r.lit(writer, indent+regexp.MustCompile(`([\[\]*/_=~+])`).ReplaceAllString(s, "$1\u200B"))
		if node.SoftLineBreak() {
			r.cr(writer)
		}
		if node.HardLineBreak() {
			r.lit(writer, `\\`)
			r.cr(writer)
		}
	}
	return ast.WalkContinue, nil
}

func (r *orgRenderer) paragraph(writer io.Writer, source []byte, n ast.Node, entering bool, rc renderer.Context) (ast.WalkStatus, error) {
	node := n.(*ast.Paragraph)
	if node.Parent().Kind() != ast.KindListItem {
		r.lit(writer, "\n")
	}
	return ast.WalkContinue, nil
}

func (r *orgRenderer) list(writer io.Writer, source []byte, n ast.Node, entering bool, rc renderer.Context) (ast.WalkStatus, error) {
	_ = n.(*ast.List)
	if entering {
		r.cr(writer)
	}
	return ast.WalkContinue, nil
}

func (r *orgRenderer) item(writer io.Writer, source []byte, n ast.Node, entering bool, rc renderer.Context) (ast.WalkStatus, error) {
	_ = n.(*ast.ListItem)
	if entering {
		if r.indent != "" {
			return ast.WalkStop, errors.New("no support for nested lists")
		}
		r.lit(writer, "- ")
		r.indent = "  "
	} else {
		if r.indent != "  " {
			return ast.WalkStop, errors.New("no support for nested lists")
		}
		r.indent = ""
		r.cr(writer)
	}
	return ast.WalkContinue, nil
}

func (r *orgRenderer) emph(writer io.Writer, source []byte, n ast.Node, entering bool, rc renderer.Context) (ast.WalkStatus, error) {
	_ = n.(*ast.Emphasis)
	r.lit(writer, "/")
	return ast.WalkContinue, nil
}

func (r *orgRenderer) code(writer io.Writer, source []byte, n ast.Node, entering bool, rc renderer.Context) (ast.WalkStatus, error) {
	node := n.(*ast.CodeSpan)
	r.lit(writer, "~")
	if entering {
		r.lit(writer, node.Value.Str(source))
	}
	return ast.WalkContinue, nil
}

func (r *orgRenderer) codeBlock(writer io.Writer, source []byte, n ast.Node, entering bool, rc renderer.Context) (ast.WalkStatus, error) {
	node := n.(*ast.CodeBlock)
	if entering {
		lang, ok := node.Language(source)
		if !ok {
			return ast.WalkStop, errors.New("language not given")
		}
		lang = rendererLanguage[lang]
		if lang == "" {
			return ast.WalkStop, fmt.Errorf("unknown language %q", lang)
		}
		r.lit(writer, fmt.Sprintf("#+BEGIN_SRC %s\n", lang))
		r.lit(writer, node.Value.Str(source))
	} else {
		r.lit(writer, "#+END_SRC\n\n")
		r.lit(writer, "#+TEXINFO: @noindent")
	}
	return ast.WalkContinue, nil
}

func (r *orgRenderer) link(writer io.Writer, source []byte, n ast.Node, entering bool, rc renderer.Context) (ast.WalkStatus, error) {
	node := n.(*ast.Link)
	if entering {
		dest := node.Destination.Str(source)
		// CommonMark helpfully URL-escapes link destinations, but this
		// prevents links to Info nodes containing spaces.
		match := regexp.MustCompile(`^(info:[^#:]+[#:])(.*%.*)$`).FindStringSubmatch(dest)
		if match != nil {
			s, err := url.PathUnescape(match[2])
			if err != nil {
				return ast.WalkStop, err
			}
			dest = match[1] + s
		}
		r.lit(writer, fmt.Sprintf("[[%s][", dest))
	} else {
		r.lit(writer, "]]")
	}
	return ast.WalkContinue, nil
}

func (r *orgRenderer) htmlInline(writer io.Writer, source []byte, n ast.Node, entering bool, rc renderer.Context) (ast.WalkStatus, error) {
	node := n.(*ast.RawHTML)
	if entering {
		tag := node.Value.Str(source)
		org := tags[tag]
		if org == "" {
			return ast.WalkStop, fmt.Errorf("unknown HTML tag %s", tag)
		}
		r.lit(writer, org)
	}
	return ast.WalkContinue, nil
}

// Signal an error if we don’t implement something.
func (r *orgRenderer) unknown(writer io.Writer, source []byte, n ast.Node, entering bool, rc renderer.Context) (ast.WalkStatus, error) {
	return ast.WalkStop, fmt.Errorf("unknown node type %q", n.Kind())
}

var tags = map[string]string{
	"<code>":  "@@texinfo:@code{@@",
	"</code>": "@@texinfo:}@@",
	"<var>":   "@@texinfo:@var{@@",
	"</var>":  "@@texinfo:}@@",
}
